### ISSUE TITLE
Failing tests under RHEL 8.1 QEMU/KVM guest

### ISSUE THREAD

**Issue by kavinda-forcepoint** [mozilla/rr#2546]
I have the following setup:
- QEMU/KVM guest: Red Hat Enterprise Linux 8.1 (x86_64)
- KVM host: Red Hat Enterprise Linux 8.1 (x86_64)
- Git repository last updated May13th, 2020

When I run `cmake --build . --target test`, 38 out of 2423 tests fail. I have attached the test output to this report.

Please let me know if there is any further information I can provide. Thanks.

[TestOutput.txt](https://github.com/mozilla/rr/files/4625544/TestOutput.txt)

**Comment by rocallahan** [comment]
Hard to say. Can you run `ctest -j8 --verbose` (in the obj directory) to collect that output and put that somewhere we can look at it?

**Comment by kavinda-forcepoint** [comment]
I have attached the output from `ctest -j8 --verbose`. This time 41 tests failed.

[VerboseTestOutput.txt](https://github.com/mozilla/rr/files/4625683/VerboseTestOutput.txt)

**Comment by rocallahan** [comment]
Can you find the `sigreturn` binary in the object directory and run it outside of rr?

**Comment by kavinda-forcepoint** [comment]
Here's the output from `sigreturn`:

```
$ ./bin/sigreturn
xmm 0 incoming: 0 0 0 0
xmm 1 incoming: 0 0 0 0
xmm 2 incoming: 0 0 0 0
xmm 3 incoming: 0 0 0 0
xmm 4 incoming: 0 0 0 0
xmm 5 incoming: 0 0 0 0
xmm 6 incoming: 0 0 0 0
xmm 7 incoming: 0 0 0 0
st 0 incoming: 0 c0000000 ffff
st 1 incoming: 0 c0000000 ffff
st 2 incoming: 0 c0000000 ffff
st 3 incoming: 0 c0000000 ffff
st 4 incoming: 0 c0000000 ffff
st 5 incoming: 0 c0000000 ffff
st 6 incoming: 0 c0000000 ffff
st 7 incoming: 0 c0000000 ffff
EXIT-SUCCESS
```

**Comment by rocallahan** [comment]
Aright, now if you `rr record sigreturn`?

**Comment by kavinda-forcepoint** [comment]
Here's the output for `rr record sigreturn`:

```
$ ../bin/rr record sigreturn
rr: Saving execution to trace directory `/home/kavinda/.local/share/rr/sigreturn-0'.
xmm 0 incoming: 0 0 0 0
FAILED: errno=0 (Success)
sigreturn: /home/kavinda/rr_root/rr/src/test/sigreturn.c:158: int main(void): Assertion `"FAILED: !" && check_cond("XMM register should have been preserved" && memcmp(xmm, xmm_good, sizeof(xmm)) == 0)' failed.
Aborted
```

**Comment by rocallahan** [comment]
Can you `rr pack` that trace and put it somewhere I can download it?

**Comment by kavinda-forcepoint** [comment]
Here's a tarball of the trace directory:
[2020-05-13-sigreturn-0.tar.gz](https://github.com/mozilla/rr/files/4625797/2020-05-13-sigreturn-0.tar.gz)

**Comment by rocallahan** [comment]
This is disturbing. It looks like XMM registers are simply not restored when running under rr.
```
  real_time:5401856.354595 global_time:302, event:`SYSCALL: rt_sigreturn' (state:EXITING_SYSCALL) tid:10877, ticks:90479
rax:0x0 rbx:0xea rcx:0xffffffffffffffff rdx:0xa rsi:0x2a7d rdi:0x2a7d rbp:0x7ffd53963e90 rsp:0x681ffdd0 r8:0x7ffd53963e90 r9:0x4015d0 r10:0x7facf6decda0 r11:0x246 r12:0x681fffa0 r13:0x7ffd53963fd0 r14:0x0 r15:0x0 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xffffffffffffffff fs_base:0x7facf783e100 gs_base:0x0 st0:0x0 st1:0x0 st2:0x0 st3:0x0 st4:0x0 st5:0x0 st6:0x0 st7:0x0 ymm0:0x0 ymm1:0x15d0260334cdf6663cccf00 ymm2:0x7ffd53963d1000007facf7012310 ymm3:0x7facf785700000007facf7012939 ymm4:0x20000000000000000 ymm5:0x800000000000000e ymm6:0x0 ymm7:0x0 ymm8:0x81f68040040200007facf741ecb0 ymm9:0x0 ymm10:0x0 ymm11:0x0 ymm12:0x0 ymm13:0x0 ymm14:0x0 ymm15:0x0
```

**Comment by rocallahan** [comment]
But apparently they are restored when not running under rr.

**Comment by rocallahan** [comment]
Is it possible for you to try an up-to-date upstream kernel, say 5.6.8, in the guest, to see if that fixes it?

**Comment by rocallahan** [comment]
No wait, something else must be going on here.

**Comment by rocallahan** [comment]
Can you check if this patch fixes the `sigreturn` test for you?
```
diff --git a/src/test/sigreturn.c b/src/test/sigreturn.c
index c810fe058..f2c55b28e 100644
--- a/src/test/sigreturn.c
+++ b/src/test/sigreturn.c
@@ -143,6 +143,8 @@ static void my_raise(int sig) {
 }
 
 int main(void) {
+  // Resolve `syscall` PLT so that doesn't interfere with our setting of XMM.
+  sys_gettid();
   init();
 
   signal(SIGUSR1, handle_usr1_xmm);
```

**Comment by rocallahan** [comment]
No, there's still something else going on here :-)

**Comment by rocallahan** [comment]
OK, this is an rr issue after all. The librrpreload build is using XMM0 in a lot of places, including the syscallbuf path that all bufferable syscalls:
```
0000000000000f50 <syscall_hook>:
     f50:       55                      push   %rbp
     f51:       41 57                   push   %r15
     f53:       41 56                   push   %r14
     f55:       53                      push   %rbx
     f56:       48 81 ec 98 00 00 00    sub    $0x98,%rsp
     f5d:       48 89 fb                mov    %rdi,%rbx
     f60:       83 3c 25 28 10 00 70    cmpl   $0x0,0x70001028
     f67:       00 
     f68:       0f 85 53 05 00 00       jne    14c1 <syscall_hook+0x571>
     f6e:       c7 04 25 28 10 00 70    movl   $0x1,0x70001028
     f75:       01 00 00 00 
     f79:       83 3d dc 80 20 00 00    cmpl   $0x0,0x2080dc(%rip)        # 20905c <buffer_enabled>
     f80:       0f 84 3b 05 00 00       je     14c1 <syscall_hook+0x571>
     f86:       8a 05 05 81 20 00       mov    0x208105(%rip),%al        # 209091 <globals+0x1>
     f8c:       84 c0                   test   %al,%al
     f8e:       0f 85 2d 05 00 00       jne    14c1 <syscall_hook+0x571>
     f94:       0f 57 c0                xorps  %xmm0,%xmm0
     f97:       0f 11 44 24 10          movups %xmm0,0x10(%rsp)
```

**Comment by rocallahan** [comment]
OTOH my home-made rr release build only has librrpreload using XMM0 at one instruction (in some buffered syscall code for clock_gettime specifically).

**Comment by rocallahan** [comment]
I guess we should to save/restore XMM0 in the syscallbuf path.

**Comment by rocallahan** [comment]
OK, pull master to get 64ba09af76528e549dc479f7d6ade1a16903e31f and try again.

**Comment by kavinda-forcepoint** [comment]
I pulled in 64ba09af76528e549dc479f7d6ade1a16903e31f and this is the result:

```
$ ./rr record sigreturn
rr: Saving execution to trace directory `/home/kavinda/.local/share/rr/sigreturn-1'.
xmm 0 incoming: 0 0 0 0
xmm 1 incoming: 0 0 0 0
xmm 2 incoming: 0 0 0 0
xmm 3 incoming: 0 0 0 0
xmm 4 incoming: 0 0 0 0
xmm 5 incoming: 0 0 0 0
xmm 6 incoming: 0 0 0 0
xmm 7 incoming: 0 0 0 0
st 0 incoming: 0 c0000000 ffff
st 1 incoming: 0 c0000000 ffff
st 2 incoming: 0 c0000000 ffff
st 3 incoming: 0 c0000000 ffff
st 4 incoming: 0 c0000000 ffff
st 5 incoming: 0 c0000000 ffff
st 6 incoming: 0 c0000000 ffff
st 7 incoming: 0 c0000000 ffff
EXIT-SUCCESS
```

**Comment by kavinda-forcepoint** [comment]
I re-ran the tests. 35 failed.

Here's the output of `cmake --build . --target test`:
[TestOutput-2.txt](https://github.com/mozilla/rr/files/4630878/TestOutput-2.txt)

Here's the output of `ctest -j 8 --verbose`:
[VerboseTestOutput-2.txt](https://github.com/mozilla/rr/files/4630879/VerboseTestOutput-2.txt)

**Comment by rocallahan** [comment]
So some progress I guess. It'll probably be a few days before I can spend more time on this.

**Comment by kavinda-forcepoint** [comment]
Sounds good, thanks. Let me know if you need any more information.

**Comment by kavinda-forcepoint** [comment]
Also, in case you weren't aware, you can get a [free Red Hat Developer Subscription](https://developers.redhat.com/articles/getting-red-hat-developer-subscription-what-rhel-users-need-know/) to run the rr tests under a RHEL 8.1 QEMU/KVM guest. In any case, I would be happy to continue assisting in whatever way I can. Thanks.

**Comment by rocallahan** [comment]
OK, I'm looking at the `invalid_jump` test. I see this log
```
(rr) rsi
main () at /home/kavinda/rr_root/rr/src/test/invalid_jump.c:15
15	  ((void (*)(void))invalid_jump_here)();
```
The test fails because we're looking for
```
expect_gdb('in main')
```
My gdb 8.3 (Fedora) says
```
0x0000000000401333 in main () at /home/roc/rr/rr/src/test/invalid_jump.c:15
15	  ((void (*)(void))invalid_jump_here)();
```
I wonder why your gdb output is different.

**Comment by rocallahan** [comment]
We are passing `-n` so it's not a `.gdbinit` issue. Mucking around in gdb it looks like whether or not gdb prints `in` is somewhat unpredictable, so I guess we should just not look for it.

**Comment by rocallahan** [comment]
https://github.com/mozilla/rr/commit/c302cfdcf5259ba5203ba08c1e4a3678e033955d fixes that.

**Comment by rocallahan** [comment]
Now I'm looking at `clone_interruption`. It fails with
```
Value 0 for parameter -g was not valid (allowed range 1-4294967295)
```
The relevant test code is
```
declare -i stop_at_event=`rr --suppress-environment-warnings dump $workdir/latest-trace | \
    python $TESTDIR/clone_interruption_finder.py`
debug clone_interruption "-g $stop_at_event"
```
so `clone_interruption_finder.py` is printing `0` I guess?

**Comment by rocallahan** [comment]
So for this one, it would be helpful if you can `rr dump /tmp/rr-test-clone_interruption-jUCovkBJB/latest-trace` (or whatever the new directory name is if you run the tests again), please.

**Comment by rocallahan** [comment]
I tried installing RHEL in a VM. I installed 8.2 because that seems to be the default. I can't find `ccache` or `capnproto` packages :-(.

**Comment by Yulmart** [comment]
> I tried installing RHEL in a VM. I installed 8.2 because that seems to be the default. I can't find `ccache` or `capnproto` packages :-(.

You might be able to install those packages if you [enable the EPEL repositories](https://fedoraproject.org/wiki/EPEL) on RHEL. We ran into a similar issue when we were packaging rr.

**Comment by kavinda-forcepoint** [comment]
> I wonder why your gdb output is different.

On my RHEL 8.1, my GDB version is: GNU gdb (GDB) Red Hat Enterprise Linux 8.2-6.el8_0

**Comment by kavinda-forcepoint** [comment]
> So for this one, it would be helpful if you can `rr dump /tmp/rr-test-clone_interruption-jUCovkBJB/latest-trace` (or whatever the new directory name is if you run the tests again), please.

Here's the output from the command you suggested:
[rr-test-clone_interruption-dump.txt](https://github.com/mozilla/rr/files/4652189/rr-test-clone_interruption-dump.txt)

**Comment by kavinda-forcepoint** [comment]
> > I tried installing RHEL in a VM. I installed 8.2 because that seems to be the default. I can't find `ccache` or `capnproto` packages :-(.
> 
> You might be able to install those packages if you [enable the EPEL repositories](https://fedoraproject.org/wiki/EPEL) on RHEL. We ran into a similar issue when we were packaging rr.

@Yulmart is correct about EPEL. Here are the exact steps I used to build `rr` on a RHEL 8.1 QEMU/KVM guest:

Install the requirements:
```
# yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
# ARCH=$( /bin/arch )
# subscription-manager repos --enable "codeready-builder-for-rhel-8-${ARCH}-rpms"
# dnf install -y ccache cmake make gcc gcc-c++ gdb libgcc libgcc.i686 glibc-devel glibc-devel.i686 libstdc++-devel libstdc++-devel.i686 python3-pexpect man-pages ninja-build capnproto capnproto-libs capnproto-devel clang++
```

Build rr:
```
$ git clone https://github.com/mozilla/rr.git
$ mkdir obj && pushd obj
$ CC=clang CXX=clang++ cmake -G Ninja ../rr
$ cmake --build .
```

Install rr:
```
# cmake --build . --target install
```

**Comment by rocallahan** [comment]
I can't seem to get `ccache`/`capnproto`:
```
[roc@localhost ~]$ sudo dnf install ccache capnproto-devel
Updating Subscription Management repositories.
Red Hat Enterprise Linux 8 for x86_64 - BaseOS (RPMs)                    7.2 kB/s | 4.1 kB     00:00    
Red Hat Enterprise Linux 8 for x86_64 - AppStream (RPMs)                  10 kB/s | 4.5 kB     00:00    
Red Hat Satellite Tools 6.5 for RHEL 8 x86_64 (RPMs)                     8.9 kB/s | 3.8 kB     00:00    
Red Hat Enterprise Linux 8 for x86_64 - Supplementary (RPMs)             8.3 kB/s | 3.8 kB     00:00    
Red Hat CodeReady Linux Builder for RHEL 8 x86_64 (RPMs)                 1.2 MB/s | 3.6 MB     00:03    
No match for argument: ccache
No match for argument: capnproto-devel
Error: Unable to find a match: ccache capnproto-devel
```

**Comment by kavinda-forcepoint** [comment]
Did you install the EPEL RPM:
```
# yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
```

`dnf` output doesn't mention EPEL, hence my question.

**Comment by rocallahan** [comment]
Ah, my bad.

**Comment by rocallahan** [comment]
Have you tried using gcc to build rr instead? I suspect building with clang is responsible for some of these failures.

**Comment by kavinda-forcepoint** [comment]
I have not. Can you use `ninja` with `gcc`?

**Comment by rocallahan** [comment]
Yes.

rr can debug clang-built stuff just fine --- e.g. Firefox and Chrome are built with clang, and a lot of people use rr to debug them. However, some of the rr tests are sensitive to the way they are compiled. We can fix that, and probably should, but it's useful to separate failures due to RHEL and failures due to clang.

**Comment by rocallahan** [comment]
I built rr with gcc in RHEL 8.2 and got these failures:
```
        120 - exec_from_other_thread (Failed)
        622 - sioc (Failed)
        623 - sioc-no-syscallbuf (Failed)
        768 - block_intr_sigchld (Timeout)
        792 - clone_interruption (Failed)
        793 - clone_interruption-no-syscallbuf (Failed)
        808 - daemon_read (Timeout)
        988 - thread_open_race (Failed)
        1336 - exec_from_other_thread-32 (Failed)
```

**Comment by rocallahan** [comment]
Picked off some easy ones: https://github.com/mozilla/rr/commit/92ea35c0bcb76f5e1322ce8f9aa9998e35284384 (sioc)
https://github.com/mozilla/rr/commit/2f98657c6490134e2142c0f392b8c3b0978f3ec4 (clone_interruption)

**Comment by rocallahan** [comment]
https://github.com/mozilla/rr/commit/2922ccbc81d0f8f948ed669d3b6c75b6febec88e (exec_from_other_thread)
That was a real rr bug that just happened to show up in this test on RHEL 8 for reasons I didn't bother to figure out...

**Comment by kavinda-forcepoint** [comment]
Thanks for the info and the updates. I will pull the latest from GitHub and try again on my side.

**Comment by kavinda-forcepoint** [comment]
With the latest version from GitHub, 6 tests fail on my RHEL 8.1 VM:
[TestOutput-3.txt](https://github.com/mozilla/rr/files/4659533/TestOutput-3.txt)

**Comment by rocallahan** [comment]
At this point I'm only seeing some intermittent failures. Latest run just
```
        2099 - mmap_shared_write_exec_race-32-no-syscallbuf (Failed)
```

**Comment by rocallahan** [comment]
You might want to look into the `early_error` failure. Looks like that's deterministic for you; I'm not able to reproduce it. Rerunning the tests with `--verbose` would help.

**Comment by rocallahan** [comment]
I've seen these timeouts quite often:
```
	768 - block_intr_sigchld (Timeout)
	808 - daemon_read (Timeout)
```
But when I run them individually they pass, and in my latest test run they passed.

**Comment by kavinda-forcepoint** [comment]
Here's a test run with `--vebose` where all 4 tests that failed were `early_error*` ones:
[VerboseTestOutput-3.txt](https://github.com/mozilla/rr/files/4665241/VerboseTestOutput-3.txt)

**Comment by rocallahan** [comment]
```
820: [FATAL /home/kavinda/rr_root/rr/src/util.cc:1468:ensure_dir() errno: EACCES] Can't write to trace directory `/tmp/rr'.
```
Any idea why rr would not be able to write to `/tmp/rr`? Are you running the tests as root?

**Comment by kavinda-forcepoint** [comment]
Ah good catch. I had run `rr record` for `pluto` as root earlier, and so /tmp/rr (where the `pluto` trace directories were stored) was only writable by root. I fixed this and am currently re-running the tests. I will post the results once complete.

**Comment by kavinda-forcepoint** [comment]
The latest results have 3 failures, none of them `early_error*` ones:
```
The following tests FAILED:
        964 - string_instructions_multiwatch (Failed)
        965 - string_instructions_multiwatch-no-syscallbuf (Failed)                                               
        2098 - mmap_shared_write_exec_race-32 (Failed)
````

Here's the output from `ctest -j8 --verbose`:
[VerboseTestOutput-4.txt](https://github.com/mozilla/rr/files/4670486/VerboseTestOutput-4.txt)

**Comment by kavinda-forcepoint** [comment]
I re-ran all the tests with `ctest -j8 --verbose` after updating my local repo to 0e9806c84b43e4f7a1491d0672d5688d94808610. All of them passed, and so I am going to close this issue. If I see consistent test failures in the future, I will open new issues for those particular tests.

Thanks for all the fixes to get the tests working in RHEL 8.1 running as a KVM guest.
